Sources/Apollo/HTTPNetworkTransport.swift

```diff
@@ -41,11 +41,24 @@ public struct GraphQLHTTPResponseError: Error, LocalizedError {
   }
 }
 
+/// Used to intercept the http requests/errors/responses. This allows modifying url requests dynamically without recreating transports
+public protocol HTTPNetworkTransportDelegate: class {
+  func intercept() -> HTTPNetworkTransportInterception
+}
+
+/// An encapsulation of state for a single request (response/error) pair. If any method returns non-null it will be used in the http pipeline
+public protocol HTTPNetworkTransportInterception {
+  func willSend(request: URLRequest) -> URLRequest?
+  func didRecieve(error: Error) -> Error?
+  func didRecieve(response: HTTPURLResponse) -> HTTPURLResponse?
+}
+
 /// A network transport that uses HTTP POST requests to send GraphQL operations to a server, and that uses `URLSession` as the networking implementation.
 public class HTTPNetworkTransport: NetworkTransport {
   let url: URL
   let session: URLSession
   let serializationFormat = JSONSerializationFormat.self
+  public weak var delegate: HTTPNetworkTransportDelegate?
```

Comment:
I might miss the usage for this, but when testing this I needed to drop `weak` keyword to make everything work. 
This keyword may be not needed here (although I might miss some context that author had when doing this)

Reply:
Delegates should be weak to avoid reference cycles. As recommended in the SwiftLint [rules](https://github.com/realm/SwiftLint/blob/83a2ca4be7fc2c1d94ae99baa9409ad9f91e84c1/Rules.md#weak-delegate).